OpenUtau/ViewModels/PianoRollViewModel.cs

```diff
@@ -112,36 +113,27 @@ public class PianoRollViewModel : ViewModelBase, ICmdSubscriber {
                 if (NotesViewModel.Part == null || NotesViewModel.Part.notes.Count == 0) {
                     return;
                 }
-                try {
-                    var project = NotesViewModel.Project;
-                    var part = NotesViewModel.Part;
-                    var tempFile = Path.Combine(PathManager.Inst.CachePath, "temp.tmp");
-                    UNote? first = null;
-                    UNote? last = null;
-                    if (NotesViewModel.Selection.IsEmpty) {
-                        first = part.notes.First();
-                        last = part.notes.Last();
-                    } else {
-                        first = NotesViewModel.Selection.FirstOrDefault();
-                        last = NotesViewModel.Selection.LastOrDefault();
-                    }
-                    var sequence = Classic.Ust.WritePlugin(project, part, first, last, tempFile, encoding: plugin.Encoding);
-                    byte[]? beforeHash = HashFile(tempFile);
-                    plugin.Run(tempFile);
-                    byte[]? afterHash = HashFile(tempFile);
-                    if (beforeHash == null || afterHash == null || Enumerable.SequenceEqual(beforeHash, afterHash)) {
-                        Log.Information("Legacy plugin temp file has not changed.");
-                        return;
-                    }
-                    Log.Information("Legacy plugin temp file has changed.");
-                    var (toRemove, toAdd) = Classic.Ust.ParsePlugin(project, part, first, last, sequence, tempFile, encoding: plugin.Encoding);
+                var part = NotesViewModel.Part;
+                UNote? first;
+                UNote? last;
+                if (NotesViewModel.Selection.IsEmpty) {
+                    first = part.notes.First();
+                    last = part.notes.Last();
+                } else {
+                    first = NotesViewModel.Selection.FirstOrDefault();
+                    last = NotesViewModel.Selection.LastOrDefault();
+                }
+                var runner = new PluginRunner(PathManager.Inst);
+                runner.OnReplaceNote += ((sender, args) => {
```

Comment:
DocManager is a Core class. Maybe you don't need OnReplaceNote and OnError events? It will avoid GC complexity around event handlers.

Reply:
Thanks.
I try to more simple method.

This was for the following reasons
1. it is the responsibility of the ViewModel to add undo/redo to plugin process.
2. DocManager has state(undoGroup), there might be a problem with parallel test.
3. To easily mocking at test.

Reply:
@stakira 
Sorry, I understand the risks of using EventHandler in WPF.
But using DocManager from PluginRunner makes it difficult to test.
I want to remove the use of EventHandler and keep DocManager replaceable.

How about a fix like this:
- Change OnReplaceNote from event to member variable (delete event += from ViewModel)
- Initialize methods use DocManager
- Test methods that use TestMethod

example:
```
// for ViewModel
public PluginRunner(PathManager path, DocManager doc) {
  this.replaceMethod = (path, toAdd, toRemove) => {
    doc.doSomething()
  }
  ....
}
// for Test
public PluginRunner(PathManager path, Action replaceMethod) {
  this.replaceMethod = replaceMethod
  ....
}
```